// web-audio-engine/lib/impl/AnalyserNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var fft = require("fourier-transform");
var util = require("../util");
var AudioNode = require("./AudioNode");
var AnalyserNodeDSP = require("./dsp/AnalyserNode");

var _require = require("../constants/ChannelCountMode"),
    MAX = _require.MAX;

var MinFFTSize = 32;
var MaxFFTSize = 32768;

var AnalyserNode = function (_AudioNode) {
  _inherits(AnalyserNode, _AudioNode);

  /**
   * @param {AudioContext} context
   */
  function AnalyserNode(context) {
    _classCallCheck(this, AnalyserNode);

    var _this = _possibleConstructorReturn(this, (AnalyserNode.__proto__ || Object.getPrototypeOf(AnalyserNode)).call(this, context, {
      inputs: [1],
      outputs: [1],
      channelCount: 1,
      channelCountMode: MAX
    }));

    _this._fftSize = 2048;
    _this._minDecibels = -100;
    _this._maxDecibels = -30;
    _this._smoothingTimeConstant = 0.8;
    _this.dspInit(_this._fftSize);
    return _this;
  }

  /**
   * @return {number}
   */


  _createClass(AnalyserNode, [{
    key: "getFftSize",
    value: function getFftSize() {
      return this._fftSize;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setFftSize",
    value: function setFftSize(value) {
      value = util.clamp(value | 0, MinFFTSize, MaxFFTSize);
      value = util.toPowerOfTwo(value, Math.ceil);
      this._fftSize = value;
      this.dspUpdateSizes(this._fftSize);
    }

    /**
     * @return {number}
     */

  }, {
    key: "getFrequencyBinCount",
    value: function getFrequencyBinCount() {
      return this._fftSize / 2;
    }

    /**
     * @return {number}
     */

  }, {
    key: "getMinDecibels",
    value: function getMinDecibels() {
      return this._minDecibels;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setMinDecibels",
    value: function setMinDecibels(value) {
      value = util.toNumber(value);
      /* istanbul ignore else */
      if (-Infinity < value && value < this._maxDecibels) {
        this._minDecibels = value;
      }
    }

    /**
     * @return {number}
     */

  }, {
    key: "getMaxDecibels",
    value: function getMaxDecibels() {
      return this._maxDecibels;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setMaxDecibels",
    value: function setMaxDecibels(value) {
      value = util.toNumber(value);
      /* istanbul ignore else */
      if (this._minDecibels < value && value < Infinity) {
        this._maxDecibels = value;
      }
    }

    /**
     * @return {number}
     */

  }, {
    key: "getSmoothingTimeConstant",
    value: function getSmoothingTimeConstant() {
      return this._smoothingTimeConstant;
    }

    /**
     * @param {number}
     */

  }, {
    key: "setSmoothingTimeConstant",
    value: function setSmoothingTimeConstant(value) {
      value = util.clamp(util.toNumber(value), 0, 1);
      this._smoothingTimeConstant = value;
    }

    /**
     * @param {Float32Array} array
     */

  }, {
    key: "getFloatFrequencyData",
    value: function getFloatFrequencyData(array) {
      var _this2 = this;

      var fftSize = this._fftSize;
      var s = this._smoothingTimeConstant;
      var len = array.length;
      var waveform = new Float32Array(fftSize);
      // 1. down-mix
      this.getFloatTimeDomainData(waveform);
      // 2. Apply Blackman window
      waveform = waveform.map(function (v, i) {
        return v * _this2._blackmanTable[i] || 0;
      });
      // 3. FFT
      var spectrum = fft(waveform);
      var numFrequencyBins = spectrum.length;
      // re-size to frequencyBinCount, then do more processing
      for (var i = 0; i < numFrequencyBins && i < len; i++) {
        var v = spectrum[i];
        // 4. Smooth over data
        this._previousSmooth[i] = s * this._previousSmooth[i] + (1 - s) * v;
        // 5. Convert to dB
        v = util.toDecibel(this._previousSmooth[i]);
        // store in array
        array[i] = Number.isFinite(v) ? v : 0;
      }
    }

    /**
     * @param {Uint8Array} array
     */

  }, {
    key: "getByteFrequencyData",
    value: function getByteFrequencyData(array) {
      var dBMin = this._minDecibels;
      var dBMax = this._maxDecibels;
      var spectrum = new Float32Array(array.length);
      this.getFloatFrequencyData(spectrum);
      for (var i = 0; i < array.length; i++) {
        array[i] = Math.round(util.normalize(spectrum[i], dBMin, dBMax) * 255);
      }
    }

    /**
     * @param {Float32Array} array
     */

  }, {
    key: "getFloatTimeDomainData",
    value: function getFloatTimeDomainData(array) {
      var len = this._timeDomainBuffer.length;
      for (var i = 0; i < array.length && i < len; i++) {
        array[i] = this._timeDomainBuffer[i] || 0;
      }
    }

    /**
     * @param {Uint8Array} array
     */

  }, {
    key: "getByteTimeDomainData",
    value: function getByteTimeDomainData(array) {
      var waveform = new Float32Array(array.length);
      this.getFloatTimeDomainData(waveform);
      for (var i = 0; i < array.length; i++) {
        array[i] = Math.round(util.normalize(waveform[i], -1, 1) * 255);
      }
    }

    /**
     * @param {number} numberOfChannels
     */

  }, {
    key: "channelDidUpdate",
    value: function channelDidUpdate(numberOfChannels) {
      this.outputs[0].setNumberOfChannels(numberOfChannels);
    }
  }]);

  return AnalyserNode;
}(AudioNode);

Object.assign(AnalyserNode.prototype, AnalyserNodeDSP);

module.exports = AnalyserNode;
