// web-audio-engine/lib/impl/BiquadFilterNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var AudioNode = require("./AudioNode");
var BiquadFilterNodeDSP = require("./dsp/BiquadFilterNode");

var _require = require("../constants/ChannelCountMode"),
    MAX = _require.MAX;

var _require2 = require("../constants/AudioParamRate"),
    CONTROL_RATE = _require2.CONTROL_RATE;

var _require3 = require("../constants/BiquadFilterType"),
    LOWPASS = _require3.LOWPASS;

var _require4 = require("../constants/BiquadFilterType"),
    HIGHPASS = _require4.HIGHPASS;

var _require5 = require("../constants/BiquadFilterType"),
    BANDPASS = _require5.BANDPASS;

var _require6 = require("../constants/BiquadFilterType"),
    LOWSHELF = _require6.LOWSHELF;

var _require7 = require("../constants/BiquadFilterType"),
    HIGHSHELF = _require7.HIGHSHELF;

var _require8 = require("../constants/BiquadFilterType"),
    PEAKING = _require8.PEAKING;

var _require9 = require("../constants/BiquadFilterType"),
    NOTCH = _require9.NOTCH;

var _require10 = require("../constants/BiquadFilterType"),
    ALLPASS = _require10.ALLPASS;

var allowedBiquadFilterTypes = [LOWPASS, HIGHPASS, BANDPASS, LOWSHELF, HIGHSHELF, PEAKING, NOTCH, ALLPASS];

var BiquadFilterNode = function (_AudioNode) {
  _inherits(BiquadFilterNode, _AudioNode);

  /**
   * @param {AudioContext} context
   */
  function BiquadFilterNode(context) {
    _classCallCheck(this, BiquadFilterNode);

    var _this = _possibleConstructorReturn(this, (BiquadFilterNode.__proto__ || Object.getPrototypeOf(BiquadFilterNode)).call(this, context, {
      inputs: [1],
      outputs: [1],
      channelCount: 2,
      channelCountMode: MAX
    }));

    _this._type = LOWPASS;
    _this._frequency = _this.addParam(CONTROL_RATE, 350);
    _this._detune = _this.addParam(CONTROL_RATE, 0);
    _this._Q = _this.addParam(CONTROL_RATE, 1);
    _this._gain = _this.addParam(CONTROL_RATE, 0);

    _this.dspInit();
    _this.dspUpdateKernel(1);
    return _this;
  }

  /**
   * @return {string}
   */


  _createClass(BiquadFilterNode, [{
    key: "getType",
    value: function getType() {
      return this._type;
    }

    /**
     * @param {string} value
     */

  }, {
    key: "setType",
    value: function setType(value) {
      /* istanbul ignore else */
      if (allowedBiquadFilterTypes.indexOf(value) !== -1) {
        this._type = value;
      }
    }

    /**
     * @return {AudioParam}
     */

  }, {
    key: "getFrequency",
    value: function getFrequency() {
      return this._frequency;
    }

    /**
     * @return {AudioParam}
     */

  }, {
    key: "getDetune",
    value: function getDetune() {
      return this._detune;
    }

    /**
     * @return {AudioParam}
     */

  }, {
    key: "getQ",
    value: function getQ() {
      return this._Q;
    }

    /**
     * @return {AudioParam}
     */

  }, {
    key: "getGain",
    value: function getGain() {
      return this._gain;
    }

    /**
     * @param {Float32Array} frequencyHz
     * @param {Float32Array} magResponse
     * @param {Float32Array} phaseResponse
     */

  }, {
    key: "getFrequencyResponse",
    value: function getFrequencyResponse(frequencyHz, magResponse, phaseResponse) {
      this.dspGetFrequencyResponse(frequencyHz, magResponse, phaseResponse);
    }

    /**
     * @param {number} numberOfChannels
     */

  }, {
    key: "channelDidUpdate",
    value: function channelDidUpdate(numberOfChannels) {
      this.dspUpdateKernel(numberOfChannels);
      this.outputs[0].setNumberOfChannels(numberOfChannels);
    }

    /**
     * @return {number}
     */

  }, {
    key: "getTailTime",
    value: function getTailTime() {
      return 0.2;
    }
  }]);

  return BiquadFilterNode;
}(AudioNode);

Object.assign(BiquadFilterNode.prototype, BiquadFilterNodeDSP);

module.exports = BiquadFilterNode;
